examples/nlp/cai/generate_rl_cai_dataset.py

```diff
@@ -0,0 +1,741 @@
+    end_name_signal = "\n"  # (model_config: chat_prompt_tokens.end_of_name. NOTE: "\x0A" is '\n' in ASCII code)
+
+    begin_signal = ""
+
+    user_message_header = begin_signal + turn_token + user_token + end_name_signal
+
+    assistant_message_header = begin_signal + turn_token + assistant_token + end_name_signal
+
+    @staticmethod
+    def _apply_header_template(system_prompt: str):
+        # header/system-message ('<extra_id_0>System\n<system_prompt>\n')
+        header = (
+            ChatPromptTemplate.system_turn_token
+            + ChatPromptTemplate.system_token
+            + ChatPromptTemplate.end_name_signal
+            + system_prompt
+            + ChatPromptTemplate.end_signal
+        )
+
+        return header
+
+    @staticmethod
+    def _apply_role_template(role: str, prompt: Optional[str] = None):
+        assert role in [ChatPromptTemplate.user_token, ChatPromptTemplate.assistant_token]
+
+        # assistant message ('<extra_id_1><role_name>\n<prompt>\n')
+        assistant_message = (
+            ChatPromptTemplate.begin_signal + ChatPromptTemplate.turn_token + role + ChatPromptTemplate.end_name_signal
+        )
+
+        if prompt is not None:
+            assistant_message += prompt + ChatPromptTemplate.end_signal
+
+        return assistant_message
+
+    @staticmethod
+    def apply_user_role_template(prompt: str):
+        assert prompt is not None and prompt != ""
+        return ChatPromptTemplate._apply_role_template(ChatPromptTemplate.user_token, prompt)
+
+    @staticmethod
+    def apply_assistant_role_template(prompt: Optional[str] = None):
+        return ChatPromptTemplate._apply_role_template(ChatPromptTemplate.assistant_token, prompt)
+
+    @staticmethod
+    def apply_prompt_template(prompt: str, system_prompt: str = ""):
+        header = ChatPromptTemplate._apply_header_template(system_prompt)
+
+        # user message ('<extra_id_1>User\n<prompt>\n')
+        user_message = ChatPromptTemplate.apply_user_role_template(prompt)
+
+        # assistant message ('<extra_id_1>Assistant\n')
+        assistant_message = ChatPromptTemplate.apply_assistant_role_template(None)
+
+        # create conversation message (full prompt)
+        conversation = header + user_message + assistant_message
+
+        return conversation
+
+    @staticmethod
+    def apply_prompt_with_response_template(prompt: str, response: str, system_prompt: str = ""):
+        prompt = ChatPromptTemplate.apply_prompt_template(prompt, system_prompt=system_prompt)
+        prompt_with_response = prompt + response + ChatPromptTemplate.end_signal + ChatPromptTemplate.turn_token
+        return prompt_with_response
+
+    @staticmethod
+    def extract_response(prompt: str):
+        response = prompt.rsplit(ChatPromptTemplate.assistant_message_header, 1)[-1]
+        response = response.strip().removesuffix(ChatPromptTemplate.turn_token).strip()
+        return response
+
+
+constitution = """
```

Comment:
Good point that we definitely need to discuss. 
As CAI paper is 1.5 years old by now [= forever in LLM developement :-)] and has no official code implementation, we agreed with @gshennvm that we will roughly follow the paper and might make some changes where we see fit. Given that the supported context size for recent models is getting increasingly larger each day, we experimented with feeding the entire constitution to the judge LLM and have it judge according to all the principles at once. Empirically we found that working quite well, as the selected most harmless and most harmful responses made perfect sense on several tens of prompts that we experimented with. 

What do you think? @terrykong @gshennvm 

P.S: there are other section which is not exactly following the paper, and we might want to change those as well if we want to follow the paper exactly.  e.g. in the paper in order to rank and choose between the responses, they use the logprob of each of the responses as the criteria for selection, whereas in our implementation we use an external judge LLM as, for instance, is being done with MT-Bench for scoring LLMs.

Reply:
I personally think it's fine to deviate from the algorithm as long as it's documented; at the very least in a comment.

I do think the constitutions should be passed into the script as opposed to being a literal within the python script. It makes customization more difficult since we are mixing "data" with code

Reply:
Absolutely agree. Will fix and make sure to properly document! 
Thanks for the feedback.

Reply:
Done. Constitution is now in a separate file and paper deviation is properly documented.